## Gather

In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
patients = pd.read_csv('data/patients.csv')
treatments = pd.read_csv('data/treatments.csv')
adverse_reactions = pd.read_csv('data/adverse_reactions.csv')

## Assess

In [3]:
patients.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2

In [4]:
treatments.head(2)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   

   hba1c_change  
0           NaN  
1          0.97

In [5]:
adverse_reactions.head(2)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name      surname                 address  \
227         228       female      Laura  Ehrlichmann      1829 Bungalow Road   
335         336       female      Lixue        Hsueh     1540 Overlook Drive   
385         386       female       Rice       Yudina      4649 Worley Avenue   
5             6         male     Rafael        Costa      1140 Willis Avenue   
407         408       female      Tegan      Johnson  2636 Benson Park Drive   

               city    state  zip_code        country  \
227           Omaha       NE   68124.0  United States   
335  Crawfordsville       IN   47933.0  United States   
385      Nellysford       VA   22958.0  United States   
5     Daytona Beach  Florida   32114.0  United States   
407   Oklahoma City       OK   73107.0  United States   

                                      contact  birthdate  weight  height   bmi  
227   402-938-2157LauraEhrlichmann@einrot.com   1/3/1942   187.9      60  36.7  
335         765-359-0147LixueHsueh@dayrep.com  3/29/1925   102.7      59  20.7  
385         RiceYudina@dayrep.com434-361-8479  2/14/1986   221.8      68  33.7  
5    386-334-5237RafaelCardosoCosta@gustr.com  8/31/1931   183.9      70  26.4  
407        405-470-5088TeganJohnson@gustr.com  1/19/1928   156.6      65  26.1

In [15]:
patients.surname.value_counts()

Doe           6
Jakobsen      3
Taylor        3
Bùi           2
Nilsen        2
             ..
Petersen      1
Uspenskaya    1
Henriksen     1
Maslov        1
Terrazas      1
Name: surname, Length: 466, dtype: int64

In [16]:
patients.address.value_counts()

123 Main Street           6
648 Old Dear Lane         2
2778 North Avenue         2
2476 Fulton Street        2
1778 Rodney Street        1
                         ..
4213 Isaacs Creek Road    1
3214 Better Street        1
3499 Baker Avenue         1
2166 Tipple Road          1
4476 Center Street        1
Name: address, Length: 483, dtype: int64

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

In [24]:
treatments_cut = pd.read_csv("data/treatments_cut.csv")
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
Add additional records to treatments dataset. Concat treatments_cut into treatments_clean to have 350 records total, not 280.

##### Code

In [25]:
treatments_clean2 = pd.concat([treatments_clean, treatments_cut])
treatments_clean2.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

##### Test

In [26]:
treatments_clean2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


In [27]:
treatments_clean2.sample(20)

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
131         søren   sørensen          -  41u - 39u         7.93       7.59   
15           enco     žibrik  55u - 68u          -         7.78       7.34   
208       katrine      lynge          -  43u - 42u         7.60       7.13   
22      beatrycze    woźniak          -  26u - 27u         7.54       7.17   
248       brianna    lansell  39u - 47u          -         9.48       9.06   
60   onyekachukwu     obinna  37u - 46u          -         7.58       7.12   
241    elisabetta    lorenzo          -  28u - 24u         7.78       7.42   
141        johana       hrdá  54u - 54u          -         7.79       7.49   
200       nicolas   ferreira  43u - 51u          -         7.99       7.72   
59      leixandre     alanis  61u - 67u          -         7.74       7.32   
75      mackenzie      mckay          -  44u - 43u         9.87       9.48   
17           emma     barlow  48u - 57u          -         7.85       7.39   
202       martina  tománková          -  50u - 57u         7.83       7.36   
48        nebechi  ekechukwu          -  37u - 39u         7.78       7.39   
54           oles    zhdanov  54u - 67u          -         7.52       7.11   
159        yunadi   barsukov  48u - 58u          -         9.47       9.05   
173      gabriele    saenger  22u - 30u          -         8.35       7.97   
69         armina      sauvé  36u - 46u          -         7.86       7.40   
22    brancaleone      russo  53u - 60u          -         8.61       8.18   
249          kang        mai          -  39u - 36u         7.78       7.45   

     hba1c_change  
131          0.34  
15            NaN  
208           NaN  
22           0.37  
248           NaN  
60            NaN  
241           NaN  
141          0.30  
200          0.27  
59            NaN  
75           0.39  
17            NaN  
202          0.97  
48           0.39  
54            NaN  
159           NaN  
173          0.38  
69            NaN  
22            NaN  
249          0.33

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
Inaccurate HbA1c values. Update all hba1c_change values by subtracting hba1c_end from hba1c_start.

##### Code

In [28]:
treatments_clean = treatments_clean2.copy()

In [29]:
treatments_clean.hba1c_change = treatments_clean.hba1c_start - treatments_clean.hba1c_end
treatments_clean.sample(10)

given_name   surname    auralin    novodra  hba1c_start  hba1c_end  \
115        xiuxiu     chang  32u - 41u          -         7.64       7.23   
173      gabriele   saenger  22u - 30u          -         8.35       7.97   
248       brianna   lansell  39u - 47u          -         9.48       9.06   
65         rovzan   kishiev  32u - 37u          -         7.75       7.41   
60   onyekachukwu    obinna  37u - 46u          -         7.58       7.12   
3             thể     lương          -  61u - 64u         7.64       7.22   
256         ilija    horvat  42u - 50u          -         7.77       7.38   
240         rocco  christie          -  42u - 40u         7.96       7.62   
5            meya   sjöberg          -  42u - 42u         7.77       7.42   
184          chân       bùi  31u - 42u          -         7.53       7.18   

     hba1c_change  
115          0.41  
173          0.38  
248          0.42  
65           0.34  
60           0.46  
3            0.42  
256          0.39  
240          0.34  
5            0.35  
184          0.35

##### Test

In [30]:
sum(treatments_clean.hba1c_change.isnull())

0

In [31]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
Separate contact data into appropriate columns, email & phone. 

Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [32]:
patients_clean.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2

In [33]:
patients_clean['email'] = patients_clean['contact'].str.extract(r'([a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z]{2,4})')
patients_clean

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  \
0    7/10/1976   121.7      66  19.6   
1     4/3/1967   118.8      66  19.2   
2    2/19/1980   177.8      71  24.8   
3    7/26/1951   220.9      70  31.7   
4    2/18/1928   192.3      27  26.1   
..         ...     ...     ...   ...   
498  4/10/1959   181.1      72  24.6   
499  3/26/1948   239.6      70  34.4   
500  1/13/1971   171.2      67  26.8   
501  2/13/1952   176.9      67  27.7   
502   5/3/1954   138.2      71  19.3   

                                           email  
0           951-719-9170ZoeWellish@superrito.com  
1                           PamelaSHill@cuvox.de  
2               402-363-6804JaeMDebord@gustr.com  
3                      PhanBaLiem@jourrapide.com  
4                334-515-7487TimNeudorf@cuvox.de  
..                                           ...  
498  207-477-0579MustafaLindstrom@jourrapide.com  
499           928-284-4492RumanBisliev@gustr.com  
500        816-223-6007JinkedeKeizer@teleworm.us  
501            ChidaluOnyekaozulu@jourrapide.com  
502                     PatrickGersten@rhyta.com  

[503 rows x 15 columns]

In [34]:
pd.isna(patients_clean.contact).value_counts()

False    491
True      12
Name: contact, dtype: int64

In [35]:
patients_clean.dropna(inplace=True)

In [36]:
pd.isna(patients_clean.contact).value_counts()

False    491
Name: contact, dtype: int64

In [37]:
def extract_phone(row):
    ph_re = re.compile(r'1?\W*([2-9][0-8][0-9])\W*([2-9][0-9]{2})\W*([0-9]{4})(\se?x?t?(\d*))?')
    found = ph_re.search(row)
    if found:
        return found.group()
    else:
        return ''

In [38]:
extract_phone(patients_clean.contact[1])

'1 (217) 569-3204'

In [39]:
patients_clean['phone'] = patients_clean.apply(lambda row: extract_phone(row['contact']), axis=1)
patients_clean

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  \
0    7/10/1976   121.7      66  19.6   
1     4/3/1967   118.8      66  19.2   
2    2/19/1980   177.8      71  24.8   
3    7/26/1951   220.9      70  31.7   
4    2/18/1928   192.3      27  26.1   
..         ...     ...     ...   ...   
498  4/10/1959   181.1      72  24.6   
499  3/26/1948   239.6      70  34.4   
500  1/13/1971   171.2      67  26.8   
501  2/13/1952   176.9      67  27.7   
502   5/3/1954   138.2      71  19.3   

                                           email             phone  
0           951-719-9170ZoeWellish@superrito.com      951-719-9170  
1                           PamelaSHill@cuvox.de  1 (217) 569-3204  
2               402-363-6804JaeMDebord@gustr.com      402-363-6804  
3                      PhanBaLiem@jourrapide.com  1 (732) 636-8246  
4                334-515-7487TimNeudorf@cuvox.de      334-515-7487  
..                                           ...               ...  
498  207-477-0579MustafaLindstrom@jourrapide.com      207-477-0579  
499           928-284-4492RumanBisliev@gustr.com      928-284-4492  
500        816-223-6007JinkedeKeizer@teleworm.us      816-223-6007  
501            ChidaluOnyekaozulu@jourrapide.com    1 360 443 2060  
502                     PatrickGersten@rhyta.com      402-848-4923  

[491 rows x 16 columns]

### Fix email column where contacts contained phone numbers BEFORE the email address. Since email can start with numbers, it is theoritcally possible to have a phone number in front of letters in an email address.

In [40]:
def remove_phone_from_email(email,phone):
    if phone in email:
        return re.sub(phone, '', email)
    else:
        return email

In [41]:
patients_clean2 = patients_clean.iloc[0:5,:].copy()
patients_clean2

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi                                 email             phone  
0  19.6  951-719-9170ZoeWellish@superrito.com      951-719-9170  
1  19.2                  PamelaSHill@cuvox.de  1 (217) 569-3204  
2  24.8      402-363-6804JaeMDebord@gustr.com      402-363-6804  
3  31.7             PhanBaLiem@jourrapide.com  1 (732) 636-8246  
4  26.1       334-515-7487TimNeudorf@cuvox.de      334-515-7487

In [42]:
patients_clean2['email']

0    951-719-9170ZoeWellish@superrito.com
1                    PamelaSHill@cuvox.de
2        402-363-6804JaeMDebord@gustr.com
3               PhanBaLiem@jourrapide.com
4         334-515-7487TimNeudorf@cuvox.de
Name: email, dtype: object

In [43]:
patients_clean2['email'] = patients_clean2.apply(lambda row: remove_phone_from_email(row['email'],row['phone']),axis=1)
patients_clean2

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi                      email             phone  
0  19.6   ZoeWellish@superrito.com      951-719-9170  
1  19.2       PamelaSHill@cuvox.de  1 (217) 569-3204  
2  24.8       JaeMDebord@gustr.com      402-363-6804  
3  31.7  PhanBaLiem@jourrapide.com  1 (732) 636-8246  
4  26.1        TimNeudorf@cuvox.de      334-515-7487

In [44]:
patients_clean['email'] = patients_clean.apply(lambda row: remove_phone_from_email(row['email'],row['phone']),axis=1)

In [45]:
patients_clean.drop(columns='contact',inplace=True)

##### Test

In [46]:
patients_clean.sample(20)

patient_id assigned_sex    given_name      surname  \
466         467       female        Andrea      Brodahl   
392         393       female       Kristin       Borgen   
443         444       female         Kajsa        Eidem   
158         159         male       Chibuzo        Okoli   
328         329       female          Anja       Hueber   
193         194         male         Borna    Šlezinger   
76           77       female        Maryam     Dratchev   
405         406         male        Marcel       Peters   
207         208       female     Beatrycze      Woźniak   
376         377       female         Ivona       Jakšić   
53           54         male  Kwemtochukwu    Ogochukwu   
498         499         male       Mustafa    Lindström   
390         391         male        Hideki    Haraguchi   
258         259         male          Abel      Yonatan   
118         119         male          Adib       Ghanem   
134         135         male         Jesse        Luoma   
349         350         male    Kristoffer    Martinsen   
388         389         male   Szilveszter        Totth   
479         480         male          Tapa  Arsanukayev   
313         314       female      Mgbankwo    Afamefuna   

                    address           city       state  zip_code  \
466     3234 Sardis Station  Golden Valley          MN   55422.0   
392  3008 Walkers Ridge Way         Lemont    Illinois   60439.0   
443    1953 Rhapsody Street          Ocala     Florida   34471.0   
158        4291 Patton Lane          Tulsa          OK   74136.0   
328    3216 Lodgeville Road    Minneapolis          MN   55402.0   
193     3073 Bedford Street     West Haven          CT    6516.0   
76        3314 Rocket Drive    Minneapolis          MN   55410.0   
405      4476 Center Street         Albany          OR   97321.0   
207        182 Cross Street        Saginaw          MI   48607.0   
376     1051 Bassell Avenue        Wichita          AR   67202.0   
53         2172 Lynn Street       Franklin          MA    2038.0   
498     2530 Victoria Court   Milton Mills          ME    3852.0   
390     3995 Colonial Drive      La Grange          TX   78945.0   
258        2621 Koontz Lane        Burbank  California   91502.0   
118      3457 Bridge Avenue      Delcambre          LA   70528.0   
134         3714 Ben Street         Albany    New York   12207.0   
349   1865 Honeysuckle Lane       Packwood          WA   98361.0   
388         2785 Norma Lane     Shreveport          LA   71109.0   
479      4720 Gordon Street        Ontario  California   91762.0   
313      1774 George Avenue         Atmore          AL   36502.0   

           country   birthdate  weight  height   bmi  \
466  United States    9/7/1994   166.1      62  30.4   
392  United States  12/12/1971   206.1      67  32.3   
443  United States   4/26/1964   191.0      60  37.3   
158  United States   6/14/1983   142.8      73  18.8   
328  United States   4/16/1987   151.8      65  25.3   
193  United States   3/21/1987   203.9      69  30.1   
76   United States   1/15/1983   190.1      63  33.7   
405  United States   2/10/1926   183.5      69  27.1   
207  United States   3/14/1946   119.2      61  22.5   
376  United States    9/5/1935   200.8      62  36.7   
53   United States   6/30/1976   150.5      72  20.4   
498  United States   4/10/1959   181.1      72  24.6   
390  United States   2/13/1997   181.3      67  28.4   
258  United States   4/29/1952   137.9      66  22.3   
118  United States  12/31/1967   254.5      72  34.5   
134  United States    6/1/1922   154.9      66  25.0   
349  United States   12/6/1964   161.5      73  21.3   
388  United States   8/14/1990   145.2      71  20.2   
479  United States   9/15/1955   220.0      65  36.6   
313  United States   7/18/1984   188.5      67  29.5   

                                email             phone  
466         AndreaBrodahl@armyspy.com  1 (612) 589-1495  
392           KristinBorgen@gustr.com      

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

In [47]:
treatments_clean.head(5)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [48]:
treat = treatments_clean.copy()

##### Define
Separate auralin & novodra columns into 3. Treatment, start dose, & end dose.

##### Code

In [49]:
val = treat.iloc[0,2]
val2 = val.split()
val2

['41u', '-', '48u']

In [50]:
def split_dose(aur_dose,nov_dose):
    if aur_dose != '-':
        val = aur_dose.split()
        start_dose = val[0]
        start_dose = re.sub('u','',start_dose)
        end_dose = val[2]
        end_dose = re.sub('u','',end_dose)
        treatment = 'auralin'
    elif nov_dose != '-':
        val = nov_dose.split()
        start_dose = val[0]
        start_dose = re.sub('u','',start_dose)
        end_dose = val[2]
        end_dose = re.sub('u','',end_dose)
        treatment = 'novodra'
    
    return pd.Series([treatment,int(start_dose),int(end_dose)])

In [51]:
treat[['treatment','start_dose','end_dose']] = treat.apply(lambda row: split_dose(row.auralin,row.novodra), axis=1)
treat.sample(5)

given_name  surname auralin    novodra  hba1c_start  hba1c_end  \
75   mackenzie    mckay       -  44u - 43u         9.87       9.48   
114       sabr    amari       -  32u - 27u         7.94       7.57   
127    farizah  sleiman       -  50u - 50u         7.80       7.40   
271        leo   vieira       -  30u - 33u         7.74       7.36   
181       kong      lei       -  32u - 30u         7.58       7.15   

     hba1c_change treatment  start_dose  end_dose  
75           0.39   novodra          44        43  
114          0.37   novodra          32        27  
127          0.40   novodra          50        50  
271          0.38   novodra          30        33  
181          0.43   novodra          32        30

##### Test

In [52]:
treat.sample(20)

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
69         ivan        fomin  25u - 32u          -         9.12       8.73   
271         leo       vieira          -  30u - 33u         7.74       7.36   
202     martina    tománková          -  50u - 57u         7.83       7.36   
197    angelico     marchesi          -  58u - 58u         7.62       7.18   
119       liisa      seppälä  36u - 47u          -         7.87       7.51   
73        mette    sandgreen  36u - 46u          -         7.56       7.08   
13       gregor         bole          -  47u - 45u         7.61       7.16   
231     kisanet     selassie  48u - 59u          -         9.64       9.27   
67        bernd    schneider  48u - 56u          -         7.74       7.44   
234      haylom        nebay          -  42u - 44u         7.62       7.22   
170         sun           ko  33u - 39u          -         7.59       7.23   
48          zak        kelly          -  38u - 38u         7.66       7.26   
276        john  teichelmann          -  49u - 49u         7.90       7.58   
264  chukwumoge    ogochukwu          -  41u - 39u         7.95       7.56   
3          skye   gormanston  33u - 36u          -         7.97       7.62   
233        arne    jørgensen  32u - 43u          -         9.65       9.31   
182       kajsa        eidem          -  41u - 38u         7.77       7.34   
244       angel        grant  31u - 37u          -         7.93       7.47   
36      chibuzo        okoli  28u - 33u          -         7.64       7.29   
94        kwasi        mensa          -  40u - 36u         8.29       7.89   

     hba1c_change treatment  start_dose  end_dose  
69           0.39   auralin          25        32  
271          0.38   novodra          30        33  
202          0.47   novodra          50        57  
197          0.44   novodra          58        58  
119          0.36   auralin          36        47  
73           0.48   auralin          36        46  
13           0.45   novodra          47        45  
231          0.37   auralin          48        59  
67           0.30   auralin          48        56  
234          0.40   novodra          42        44  
170          0.36   auralin          33        39  
48           0.40   novodra          38        38  
276          0.32   novodra          49        49  
264          0.39   novodra          41        39  
3            0.35   auralin          33        36  
233          0.34   auralin          32        43  
182          0.43   novodra          41        38  
244          0.46   auralin          31        37  
36           0.35   auralin          28        33  
94           0.40   novodra          40        36

#### Adverse reaction should be part of the `treatments` table

##### Define
Consolidate files/dataframes table, adverse_reactions. Add data to treatments dataframe. 

##### Code

In [70]:
adverse_reactions.sample(5)

given_name     surname           adverse_reaction
22      lixue       hsueh  injection site discomfort
25     elliot  richardson               hypoglycemia
10      tegan     johnson                   headache
12  abdul-nur         isa               hypoglycemia
31     steven         roy                   headache

In [59]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [58]:
treat.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
 7   treatment     350 non-null    object 
 8   start_dose    350 non-null    int64  
 9   end_dose      350 non-null    int64  
dtypes: float64(3), int64(2), object(5)
memory usage: 40.1+ KB


In [56]:
treat_adver = pd.concat([treat,adverse_reactions], join='outer')
treat_adver

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0    veronika    jindrová  41u - 48u          -         7.63       7.20   
1      elliot  richardson          -  40u - 45u         7.56       7.09   
2    yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3        skye  gormanston  33u - 36u          -         7.97       7.62   
4      alissa      montez          -  33u - 29u         7.78       7.46   
..        ...         ...        ...        ...          ...        ...   
29       alex    crawford        NaN        NaN          NaN        NaN   
30     monika      lončar        NaN        NaN          NaN        NaN   
31     steven         roy        NaN        NaN          NaN        NaN   
32    cecilie      nilsen        NaN        NaN          NaN        NaN   
33  krisztina      magyar        NaN        NaN          NaN        NaN   

    hba1c_change treatment  start_dose  end_dose adverse_reaction  
0           0.43   auralin        41.0      48.0              NaN  
1           0.47   novodra        40.0      45.0              NaN  
2           0.43   novodra        39.0      36.0              NaN  
3           0.35   auralin        33.0      36.0              NaN  
4           0.32   novodra        33.0      29.0              NaN  
..           ...       ...         ...       ...              ...  
29           NaN       NaN         NaN       NaN     hypoglycemia  
30           NaN       NaN         NaN       NaN     hypoglycemia  
31           NaN       NaN         NaN       NaN         headache  
32           NaN       NaN         NaN       NaN     hypoglycemia  
33           NaN       NaN         NaN       NaN     hypoglycemia  

[384 rows x 11 columns]

In [64]:
treatment_adverse = pd.merge(adverse_reactions,treat, on='surname', how='outer')
treatment_adverse

given_name_x     surname           adverse_reaction given_name_y  \
0          berta  napolitani  injection site discomfort        berta   
1           lena        baer               hypoglycemia         lena   
2         joseph         day               hypoglycemia       joseph   
3         joseph         day               hypoglycemia       joseph   
4         flavia  fiorentino                      cough       flavia   
..           ...         ...                        ...          ...   
347          NaN    fakhoury                        NaN       fakhri   
348          NaN         rap                        NaN        žarka   
349          NaN     kishiev                        NaN       rovzan   
350          NaN   schneider                        NaN        bernd   
351          NaN       sauvé                        NaN       armina   

       auralin    novodra  hba1c_start  hba1c_end  hba1c_change treatment  \
0            -  42u - 44u         7.68       7.21          0.47   novodra   
1            -  41u - 38u         7.70       7.40          0.30   novodra   
2    29u - 36u          -         7.70       7.19          0.51   auralin   
3    29u - 36u          -         7.70       7.19          0.51   auralin   
4    46u - 57u          -         7.95       7.60          0.35   auralin   
..         ...        ...          ...        ...           ...       ...   
347  39u - 50u          -         7.83       7.39          0.44   auralin   
348  35u - 48u          -         7.54       7.15          0.39   auralin   
349  32u - 37u          -         7.75       7.41          0.34   auralin   
350  48u - 56u          -         7.74       7.44          0.30   auralin   
351  36u - 46u          -         7.86       7.40          0.46   auralin   

     start_dose  end_dose  
0            42        44  
1            41        38  
2            29        36  
3            29        36  
4            46        57  
..          ...       ...  
347          39        50  
348          35        48  
349          32        37  
350          48        56  
351          36        46  

[352 rows x 12 columns]

In [66]:
treatment_adverse.query("surname == 'day'")

given_name_x surname adverse_reaction given_name_y    auralin novodra  \
2       joseph     day     hypoglycemia       joseph  29u - 36u       -   
3       joseph     day     hypoglycemia       joseph  29u - 36u       -   

   hba1c_start  hba1c_end  hba1c_change treatment  start_dose  end_dose  
2          7.7       7.19          0.51   auralin          29        36  
3          7.7       7.19          0.51   auralin          29        36

In [68]:
dups = treatment_adverse[treatment_adverse['surname'].duplicated()]

In [69]:
dups

given_name_x    surname adverse_reaction given_name_y    auralin  \
3         joseph        day     hypoglycemia       joseph  29u - 36u   
8         louise    johnson     hypoglycemia        tegan          -   
9          tegan    johnson         headache       louise  32u - 42u   
10         tegan    johnson         headache        tegan          -   
13           noe     aranda     hypoglycemia          noe  26u - 34u   
21         jakob   jakobsen     hypoglycemia        jakob          -   
38       cecilie     nilsen     hypoglycemia      bjørnar          -   
48           NaN    woźniak              NaN    beatrycze          -   
63           NaN       berg              NaN      novalie          -   
75           NaN        lâm              NaN        hường  40u - 49u   
139          NaN      lương              NaN          thể          -   
164          NaN  ogochukwu              NaN   chukwumoge          -   
180          NaN    cindrić              NaN     bernarda  40u - 49u   
191          NaN     tucker              NaN       joseph  48u - 56u   
199          NaN   dratchev              NaN       maryam  37u - 45u   
214          NaN        bùi              NaN         diệt          -   
224          NaN  schiavone              NaN       albino  35u - 43u   
241          NaN     hueber              NaN         anja          -   
253          NaN  batukayev              NaN      satsita          -   

       novodra  hba1c_start  hba1c_end  hba1c_change treatment  start_dose  \
3            -         7.70       7.19          0.51   auralin          29   
8    34u - 34u         7.79       7.43          0.36   novodra          34   
9            -         7.63       7.32          0.31   auralin          32   
10   34u - 34u         7.79       7.43          0.36   novodra          34   
13           -         7.51       7.17          0.34   auralin          26   
21   28u - 26u         7.96       7.51          0.45   novodra          28   
38   36u - 33u         7.99       7.70          0.29   novodra          36   
48   26u - 27u         7.54       7.17          0.37   novodra          26   
63   32u - 31u         7.85       7.49          0.36   novodra          32   
75           -         7.66       7.29          0.37   auralin          40   
139  61u - 64u         7.64       7.22          0.42   novodra          61   
164  41u - 39u         7.95       7.56          0.39   novodra          41   
180          -         7.89       7.55          0.34   auralin          40   
191          -         7.67       7.30          0.37   auralin          48   
199          -         9.91       9.46          0.45   auralin          37   
214  30u - 32u         9.11       8.76          0.35   novodra          30   
224          -         7.56       7.15          0.41   auralin          35   
241  31u - 32u         8.82       8.36          0.46   novodra          31   
253  42u - 42u         7.63       7.25          0.38   novodra          42   

     end_dose  
3          36  
8          34  
9          42  
10         34  
13         34  
21         26  
38         33  
48         27  
63         31  
75         49  
139        64  
164        39  
180        49  
191        56  
199        45  
214        32  
224        43  
241        32  
253        42

##### Test

In [ ]:
# Your testing code here

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here